### ISSUE TITLE
[New build - DEFAULT] 24w018a: Docking Connector

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#5733]
The docking connector have been given some love.

- New texture. The docking status now shows up as a glowing color.
- Can initiate a docking manually by using a button when near another dock. 
The connector will aling the entity and complete the docking procedure automatically.
- Can break the docking manually too.
- New "Is docking available" event, that trigger when a docking procedure can be started.
- New docking alignment setting: locked, 90 degree lock or free. Adjust whether and how the entity should align.
- Connect to a screen to view the different status: Off, Docking available, Docking Proceeding, Docked.

Hotfixes:
#5728 Hair still present when changing character from human to alien
#5726 Crash when building on a moving platform
#5719 Prefab and item stands

**Comment by COMICLER** [comment]
FINALLY. ive been waiting for this,awesome

**Comment by king32109** [comment]
it docking connector does act strange tho, i would have expected to have the entitie docked to it to be static and stay on the docking connector, and not like its some kind of rubber band effect where you move your ship and then the docked ship slowly rubber bands back to the docking port, would like to see that changed so the docked ship stays put on the docking conector and not free move on it

**Comment by TIKIRobo** [comment]
https://github.com/tsunamayo/Starship-EVO/assets/56371294/15d0eae4-b44a-4789-8201-c1dde11a4f98

very neat! however two major issues, first docking ports are inconsistent and dont always actually dock and second as shown in the next video switches kinda freak out and have to be toggled twice  


https://github.com/tsunamayo/Starship-EVO/assets/56371294/af6244b8-c896-4408-9266-e2fdeffafba6

(also an example of something failing to properly connect)
I believe that things fail to connect because if the larger entity moves towards the smaller one its larger mass pushes the smaller one away before they can lock

**Comment by COMICLER** [comment]
Im also noticing that it feels practically impossible to dock unless your in a very controlled environment, i think it would help if once you got close enough ,maybe around 1/8th of a meter they just snap together

**Comment by tsunamayo** [comment]
Okay I docked with a static entity in my test, I will try with a non static one too. @COMICLER There is a threshold already, more like 3cm for now (otherwise it create a visual snap).

**Comment by king32109** [comment]
Also noticed it takes a very long time to dock before it turns green and is rather inconsistent on it

**Comment by tsunamayo** [comment]
@king32109 Are you docking to a static station?

**Comment by tsunamayo** [comment]
I forgot to add in the description, docking port now can connect to pipe/conveyor and share inventory.

**Comment by king32109** [comment]
I docked to a moving grid

**Comment by tsunamayo** [comment]
@king32109 well then same issue. It must be pushing the other entity, which makes the docking harder to complete.

**Comment by tsunamayo** [comment]
Can a kind soul send me a world with one of your fiddly test? I had no trouble even with two non static entities.

**Comment by ProPeach** [comment]
@tsunamayo here's just a minimum test, even when the connectors line up okay-ish they have trouble finalising the docking it seems - kinda seems like they have the most trouble rotating the ship to the correct orientation.
https://i.imgur.com/wE8Tfwq.mp4
[img]https://i.imgur.com/hZf2Fbn.mp4[/img]

[Docking Test.zip](https://github.com/tsunamayo/Starship-EVO/files/15160298/Docking.Test.zip)

**Comment by ProPeach** [comment]
The issue of ships moving away rather than docking also has been an issue with Shipyards if I remember correctly. One way this could be helped is having ships "on" even while the pilot is not controlling their yoke, so they apply a counter thrust in an effort to stay in the same position and resist being moved by a docking ship. This would also allow you to keep your ship flying in gravity without it just dropping, same for hovercraft. It's a feature of ships in Space Engineers for example - Exiting control of the ship does not turn off its thrusters, that's a separate action

**Comment by Crimson-Artist** [comment]
@tsunamayo it says that we can now transfer liquids and items between entities. How do we do that?

**Comment by tsunamayo** [comment]
@drimmel247 Well the connectors already works like hoses, you can transfer item / liquid.

**Comment by tsunamayo** [comment]
> @tsunamayo it says that we can now transfer liquids and items between entities. How do we do that?

You connect a pipe/tube to the connector. Then from your storage container you do the transfer like always, you will see the inventory of the other entity inside.

**Comment by tsunamayo** [comment]
> The issue of ships moving away rather than docking also has been an issue with Shipyards if I remember correctly. One way this could be helped is having ships "on" even while the pilot is not controlling their yoke, so they apply a counter thrust in an effort to stay in the same position and resist being moved by a docking ship. This would also allow you to keep your ship flying in gravity without it just dropping, same for hovercraft. It's a feature of ships in Space Engineers for example - Exiting control of the ship does not turn off its thrusters, that's a separate action. It's kinda tied to their "inerial dampeners" mechanic, if the dampeners are on then the ship resists movement execpt commands given by the player pressing WASD (like sevo 6dof mode), if the dampeners are off then the ship can be pushed about or fall in gravity.
> 
> Like here - The ship is pushing down on the pad, and the pad uses its thrusters to resist the ship and stay in position rather than float away
> 
>  SE.Landing.Dampeners.mp4

Thanks for the insight, this was the solution I planned to do, right now I only apply force on the docking entity, but I planned to add a corrective force on the mother entity to cancel its velocity.

**Comment by ProPeach** [comment]
Oh great, glad to see you're on it then 👌

**Comment by Crimson-Artist** [comment]
> You connect a pipe/tube to the connector. Then from your storage container you do the transfer like always, you will see the inventory of the other entity inside.
@tsunamayo I tried doing that. It doesn't seem to work. I Have both docking connectors hooked up to cargo containers via conveyor tubes. upon docking I can select the other cargo on the other ship however the items inside the first cargo are not moving over. I've also noticed that the transfer window doesn't save its target. I keep setting it for the other container but it resets the moment I close the window.

**Comment by tsunamayo** [comment]
@Crimson-Artist I dont think it is supposed to keep the previous setting when you open again, that is just a UI setting. But the non transfer seems a legit bug, I will check it out.

**Comment by Horkas** [comment]
@tsunamayo Do not be afraid to have some visual snap going on, if it means more reliability and ease of use for the user to dock. I can 100% assure you, 99.9% of your players will not care about it when they are busy having fun with the rest of the game. Do what works most reliably.

**Comment by tsunamayo** [comment]
@Crimson-Artist Hi, I didnt any trouble to transfert items...
